CRM/Core/Transaction.php

```diff
@@ -248,6 +248,7 @@ static public function isActive() {
    */
   static public function addCallback($phase, $callback, $params = NULL, $id = NULL) {
     $frame = \Civi\Core\Transaction\Manager::singleton()->getBaseFrame();
+    if (empty($frame)) return NULL;
```

Comment:
I spoke a little loosely with "the caller expects the callback to run, but it will never run". I still believe this but should offer more detail.

Suppose a caller says this:

```php
DB::executeQuery('UPDATE foo');
CRM_Core_Transaction::addCallback(
  CRM_Core_Transaction::PHASE_POST_COMMIT, 
  'updateOtherData'
);
```

The intent is that 'updateOtherData' will run after 'UPDATE foo' has been fully written to the DB. In the case where a transaction is active, it does exactly that. In the case where there's no transaction, 'UPDATE foo' writes immediately (because there's no TX) but the calllback 'updateOtherData' never runs (because there's no TX and no COMMIT). That's a buggy situation. It is an error to use CRM_Core_Transaction::addCallback() if there is no transaction.